src/main/java/de/maxhenkel/audioplayer/command/AudioPlayerCommands.java

```diff
@@ -401,26 +401,26 @@ private static LiteralArgumentBuilder<CommandSourceStack> bulkApplyCommand(Liter
                                     UUID sound = UuidArgument.getUuid(context, "sound");
                                     ItemStack itemInHand = player.getItemInHand(InteractionHand.MAIN_HAND);
                                     String customName = StringArgumentType.getString(context, "custom_name");
-                                    if (validator.test(itemInHand)) {
-                                        processShulker(context, itemInHand, actionItem, itemTypeName, sound, customName);
+                                    if (containerValidator.test(itemInHand)) {
+                                        processShulker(context, itemInHand, itemValidator, itemTypeName, sound, customName);
                                     } else {
                                         context.getSource().sendFailure(Component.literal("You don't have a %s in your main hand".formatted(itemTypeName)));
                                     }
                                     return 1;
                                 })));
     }
 
-    private static void processShulker(CommandContext<CommandSourceStack> context, ItemStack itemInHand, String actionItem, String itemTypeName, UUID soundID, @Nullable String name) {
+    private static void processShulker(CommandContext<CommandSourceStack> context, ItemStack itemInHand, Predicate<ItemStack> itemValidator, String itemTypeName, UUID soundID, @Nullable String name) {
         ListTag shulkerContents = Objects.requireNonNull(itemInHand.getTagElement(BlockItem.BLOCK_ENTITY_TAG)).getList(ShulkerBoxBlockEntity.ITEMS_TAG, Tag.TAG_COMPOUND);
         for (int i = 0; i < shulkerContents.size(); i++) {
             CompoundTag currentItem = shulkerContents.getCompound(i);
             ItemStack itemStack = ItemStack.of(currentItem);
-            if ((itemStack.getItem() instanceof RecordItem && actionItem.equals("musicdisc")) || (itemStack.getItem() instanceof InstrumentItem && actionItem.equals("goathorn"))) {
+            if (itemValidator.test(itemStack)) {
                 renameItem(context, itemStack, soundID, name);
                 shulkerContents.getCompound(i).put("tag", itemStack.getTag());
```

Comment:
I don't think this is the cleanest way to change the item NBT. You should use `ItemStack.save(tag)`

Reply:
This was the toughest issue to solve, going from ItemStack type to Tag type. I'm not trying to update the ItemStack here but rather extract the updated "tag" data and overwrite only that piece in the CompoundTag.

Reply:
The issue is that the itemstack does not have all the necessary data to create a new compound. `currentItem` has data like Count, Slot, id, optionally "tag", and it could possibly have unknown data from other mods. ItemStack does not preserve the Slot data because it's no longer in the context of being in a shulker box.

My approach instead was to perform custom tagging operations on the ItemStack, extract the completed "tag", and then overwrite the `currentItem` "tag" without touching or losing anything like Slot, Count, id, etc.

(I noticed I can change line 420 to `currentItem.put("tag", itemStack.getTag());` since I saved a variable for that purpose, will update)